### ISSUE TITLE
twisted.web.test.requesthelper.DummyRequest is out of sync with the real Request

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8140]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#8140|
|Type|defect|
|Created|2015-12-06 07:56:10Z|
|Branch|https://github.com/twisted/twisted/tree/dummyrequest-update-8140|

How it handles headers is out of sync -- nothing should use ```.headers``` now, and it has been removed from Twisted. DummyRequest should be updated to match.

Attachments:

* [twisted_web_test_DummyRequest-8140.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/32d/32dca7eeea8e8a81f732dc6b07f10566850c03b5/e72cbc16aae83dae9e53a85418d43830f586c032.patch) (27409 bytes) - added by jrolli on 2016-01-01 08:20:40Z - 
* [twisted_web_test_DummyRequest-8140-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/32d/32dca7eeea8e8a81f732dc6b07f10566850c03b5/268a06c1736338412c26e8b10833a913778f8f64.patch) (4889 bytes) - added by jrolli on 2016-01-04 16:42:08Z - Addresses adiroiban's comments
* [twisted_web_test_DummyRequest-8140-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/32d/32dca7eeea8e8a81f732dc6b07f10566850c03b5/cb97ddc74244f68241539e3ec8f6b656bea69ac3.patch) (1409 bytes) - added by jrolli on 2016-01-06 17:01:10Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__8140 8140
type__defect defect
reporter__hawkowl hawkowl
priority__normal normal
milestone__Twisted_16_0 Twisted 16.0
branch__branches_dummyrequest_update_8140 branches/dummyrequest-update-8140
branch_author__adiroiban__jrolli adiroiban, jrolli
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1449388570434052 1449388570434052
changetime__1452315333146848 1452315333146848
version__None None
owner__jrolli jrolli

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="jrolli's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|jrolli commented|
|-|-|

Patch removes all references on the attributes ```headers``` and ```outgoingHeaders```.  ```DummyRequest``` now uses the Header objects ```responseHeaders``` and ```requestHeaders```

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban commented|
|-|-|

(In [46582]) Branching to dummyrequest-update-8140.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to jrolli|
|-|-|

Many thanks for you contribution!

-----

# Major comments

Why do we still need the ``` DummyRequest._headers ``` ?

Do we still need DummyRequest.outgoingHeaders ?

-------

You also need to update the docstring of DummyRequest

--------

This needs a bit of cleanup

twisted/web/test/test_static.py:22: 'iteritems' imported but unused

-------

This test needs to be fixed

```

Traceback (most recent call last):
  File "/buildslave/debian7-python-3.4-tests/Twisted/twisted/web/test/test_static.py", line 718, in test_multipleUnsatisfiableRangeSetsContentHeaders
    resource.makeProducer(request, file)
  File "/buildslave/debian7-python-3.4-tests/Twisted/twisted/web/static.py", line 551, in makeProducer
    self._setContentHeaders(request, size)
  File "/buildslave/debian7-python-3.4-tests/Twisted/twisted/web/static.py", line 519, in _setContentHeaders
    request.setHeader(b'content-type', networkString(self.type))
  File "/buildslave/debian7-python-3.4-tests/Twisted/twisted/python/compat.py", line 443, in networkString
    raise TypeError("Can only convert text to bytes on Python 3")
builtins.TypeError: Can only convert text to bytes on Python 3

twisted.web.test.test_static.StaticMakeProducerTests.test_multipleUnsatisfiableRangeSetsContentHeaders


```

------

I think that the docstring of DummyRequest.getAllHeaders should contain a note that it is a copy/paste from twisted.web.http.Request or maybe we can use this implementation

```
    def getAllHeaders(self):
        """
        Dirty hack to avoid code duplication.
        """
        return Request.getAllHeaders(self)
```



----

# Minor commets

This is stragen indentation :)

```
+        self.request.requestHeaders.addRawHeader(b'referer',
+                               b'http://malicious" ".website.invalid')
```


-----

Please review my comments and resubmit a new patch, preferably on top of the current branch.

Thanks again!

**Comment by twisted-trac** [comment]
|<img alt="jrolli's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|jrolli removed owner|
|-|-|

Apply patch to branch ```dummyrequest-update-8140```.

This addresses all of adiroiban's concerns.  Best solution I could come up with that avoids code duplication was to add ```self.getAllHeaders = Request.getAllHeaders``` to ```DummyRequest```'s ```__init__}}.  This should reuse all of the original code to include the docstring```{__init__}}.  This should reuse all of the original code to include the docstring.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban set owner to jrolli|
|-|-|

tests are failing ... due to ``` self.getAllHeaders = Request.getAllHeaders ```


```
[ERROR]
Traceback (most recent call last):
  File "C:\Users\Build Slave\bot-glyph-6\windows7-64-py2.7-coverage\Twisted\twisted\web\test\test_cgi.py", line 306, in test_pathInfo
    _render(resource, request)
  File "C:\Users\Build Slave\bot-glyph-6\windows7-64-py2.7-coverage\Twisted\twisted\web\test\_util.py", line 21, in _render
    result = resource.render(request)
  File "C:\Users\Build Slave\bot-glyph-6\windows7-64-py2.7-coverage\Twisted\twisted\web\twcgi.py", line 117, in render
    for title, header in request.getAllHeaders().items():
exceptions.TypeError: unbound method getAllHeaders() must be called with Request instance as first argument (got nothing instead)

twisted.web.test.test_cgi.CGIScriptTests.test_pathInfo
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "C:\Users\Build Slave\bot-glyph-6\windows7-64-py2.7-coverage\Twisted\twisted\web\test\test_cgi.py", line 271, in test_useReactorArgument
    _render(resource, request)
  File "C:\Users\Build Slave\bot-glyph-6\windows7-64-py2.7-coverage\Twisted\twisted\web\test\_util.py", line 21, in _render
    result = resource.render(request)
  File "C:\Users\Build Slave\bot-glyph-6\windows7-64-py2.7-coverage\Twisted\twisted\web\twcgi.py", line 117, in render
    for title, header in request.getAllHeaders().items():
exceptions.TypeError: unbound method getAllHeaders() must be called with Request instance as first argument (got nothing instead)

twisted.web.test.test_cgi.CGITests.test_useReactorArgument
```

you might want to define it at "compile time" and not at run time

-------

you should be able to run the test on your computer using

```
./bin/trial twisted.web
```

-----

topfile NEWS fragments are designed for public "end user" announcements.

There is not much value to inform end users about each changes into the private things.

the .misc is there just to acknowledge that this is not a public thing.

----

Please look at sending a new patch on top of latest branch.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="jrolli's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|jrolli removed owner|
|-|-|

adiroiban,\\
Sorry about the test regression.  I got fixated on solving errors in one module and forgot to back out before submitting.  Latest patch passes all tests in Python 2.7 and all the tests trunk does in Python 3.4.
----

Apply to branch ```dummyrequest-update-8140```.

Removed topfile NEWS fragment.

Replaced attempt at code reuse with copy-paste from ```twisted.web.http.Request.getAllHeaders``` and added a note in the docstring stating that it is a direct copy.  Could not find an elegant solution around ```Request.getAllHeaders``` expecting ```self``` to have type ```Request```.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban set owner to jrolli|
|-|-|

Thanks. Applied your patch and send it to buildbot.

If all is ok I will merge.

Many thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl commented|
|-|-|

There was a small PyFlakes warning (unused import), which I've fixed, and set it to the buildbots.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl set status to `closed`|
|-|-|

(In [46591]) Merge dummyrequest-update-8140: twisted.web.test.requesthelper.DummyRequest synchronised with the real Request

Author: jrolli
Reviewer: adiroiban
Fixes: [#8140](https://github.com/twisted/twisted/issues/8140)
